File: samples/greedy-c/strategy.c
/* Greedy strategy, kept out of bot.c so bot.c stays a thin wrapper. Every
 * *.c file in this folder is compiled in automatically (see the template
 * Makefile's `wildcard *.c`), so no build config changes needed. */
#include <string.h>

#include "strategy.h"

static void drop(const int board[COLS][ROWS], int col, int piece, int out[COLS][ROWS]) {
    memcpy(out, board, sizeof(int) * COLS * ROWS);
    for (int r = 0; r < ROWS; r++) {
        if (out[col][r] == 0) {
            out[col][r] = piece;
            break;
        }
    }
}

static int wins(const int board[COLS][ROWS], int piece) {
    static const int dirs[4][2] = { { 1, 0 }, { 0, 1 }, { 1, 1 }, { 1, -1 } };
    for (int col = 0; col < COLS; col++) {
        for (int row = 0; row < ROWS; row++) {
            if (board[col][row] != piece) continue;
            for (int d = 0; d < 4; d++) {
                int count = 1;
                int c = col + dirs[d][0];
                int r = row + dirs[d][1];
                while (c >= 0 && c < COLS && r >= 0 && r < ROWS && board[c][r] == piece) {
                    count++;
                    c += dirs[d][0];
                    r += dirs[d][1];
                }
                if (count >= 4) return 1;
            }
        }
    }
    return 0;
}

int greedy_move(const int board[COLS][ROWS], int you) {
    int opponent = you == 1 ? 2 : 1;
    int moves[COLS];
    int n = legal_moves(board, moves);
    int tmp[COLS][ROWS];

    for (int i = 0; i < n; i++) {
        drop(board, moves[i], you, tmp);
        if (wins(tmp, you)) return moves[i];
    }
    for (int i = 0; i < n; i++) {
        drop(board, moves[i], opponent, tmp);
        if (wins(tmp, opponent)) return moves[i];
    }
    double center = (COLS - 1) / 2.0;
    int best = moves[0];
    for (int i = 0; i < n; i++) {
        double a = moves[i] - center;
        double b = best - center;
        if ((a < 0 ? -a : a) < (b < 0 ? -b : b)) best = moves[i];
    }
    return best;
}

File: samples/greedy-java/Strategy.java
import java.util.ArrayList;
import java.util.List;

// Greedy strategy, kept out of Bot.java so Bot.java stays a thin wrapper.
// javac compiles every *.java in the build context, so this file is picked
// up automatically (no Dockerfile change needed).
public class Strategy {
    private static final int[][] DIRS = { { 1, 0 }, { 0, 1 }, { 1, 1 }, { 1, -1 } };

    public static List<Integer> legalMoves(int[][] board) {
        List<Integer> moves = new ArrayList<>();
        for (int col = 0; col < board.length; col++) {
            if (board[col][board[col].length - 1] == 0) {
                moves.add(col);
            }
        }
        return moves;
    }

    private static int[][] drop(int[][] board, int col, int piece) {
        int[][] next = new int[board.length][];
        for (int c = 0; c < board.length; c++) {
            next[c] = board[c].clone();
        }
        for (int r = 0; r < next[col].length; r++) {
            if (next[col][r] == 0) {
                next[col][r] = piece;
                break;
            }
        }
        return next;
    }

    private static boolean wins(int[][] board, int piece) {
        int w = board.length, h = board[0].length;
        for (int col = 0; col < w; col++) {
            for (int row = 0; row < h; row++) {
                if (board[col][row] != piece) continue;
                for (int[] d : DIRS) {
                    int count = 1;
                    int c = col + d[0], r = row + d[1];
                    while (c >= 0 && c < w && r >= 0 && r < h && board[c][r] == piece) {
                        count++;
                        c += d[0];
                        r += d[1];
                    }
                    if (count >= 4) return true;
                }
            }
        }
        return false;
    }

    public static int greedyMove(int[][] board, int you) {
        int opponent = you == 1 ? 2 : 1;
        List<Integer> moves = legalMoves(board);
        double center = (board.length - 1) / 2.0;

        for (int col : moves) {
            if (wins(drop(board, col, you), you)) return col;
        }
        for (int col : moves) {
            if (wins(drop(board, col, opponent), opponent)) return col;
        }
        int best = moves.get(0);
        for (int col : moves) {
            if (Math.abs(col - center) < Math.abs(best - center)) best = col;
        }
        return best;
    }
}
